### PR TITLE
filter anyOf and oneOf alternative

### LINKED ISSUE #175: Filter anyOf/oneOf object alternative results by the discriminator
Let's start with an example.

```json
{
  "anyOf": [
    {
      "type": "object",
      "properties": {
        "type": { "const": "a" },
        "a": { "type": "string" }
      },
      "required": ["type", "a"]
    },
    {
      "type": "object",
      "properties": {
        "type": { "const": "b" },
        "b": { "type": "number" }
      },
      "required": ["type", "b"]
    }
  ]
}
```

If given `{}`, we don't know which alternative was intended, so we give the errors for both alternatives.

- Expected the value to match at least one alternative
  - Alternative 1
    - Missing required properties: "type" or "a"
  - Alternative 2
    - Missing required properties: "type" or "a"

If given `{ "type": "a" }`, we know that the the user intended the first alternative and can collapse the message to the messages for that one schema.

- Missing required property: "a"

In this case, it's clear that "type" is the discriminator and if "type" matches the `const` keyword, that's the alternative that's intended. But, there are other more subtle discriminators. The property name can be anything, the schema of a discriminator can be anything, and the discriminator can be more than one property.

We'll define the discriminator as the set of properties defined by all object alternatives. If an object alternative passed all the discriminator property(s), then any other alternatives that don't match the discriminator can be filtered out.

As before if there is only one alternative left, it can be collapsed and the generic `anyOf`/`oneOf` message doesn't need to be returned.

Keep in mind that an `anyOf`/`oneOf` can be have both objects and non-objects and the discriminator can still be used to filter the object alternatives.

### PR DESCRIPTION
**Pull request by animeshsahoo1**
The anyof and oneof error handlers now filter alternatives based on the instance's properties. For object instances, an alternative is excluded from the error report if none of its declared properties exist on the instance (Rule 1), or if none of the instance's properties pass validation in that alternative (Rule 2). For non-object instances, the original type-keyword filtering is done.

fixes #175

### CODE REVIEW

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
Please add tests. They need to cover all the edge cases we discussed and more if you can think of them.

File: src/error-handlers/anyOf.js [hunk 1]
```diff
@@ -2,45 +2,88 @@ import * as Instance from "@hyperjump/json-schema/instance/experimental";
...
 
+/** @type (alternative: NormalizedOutput, propLocation: string) => boolean */
+const propertyPasses = (alternative, propLocation) => {
+  const propOutput = alternative[propLocation];
+  if (!propOutput || Object.keys(propOutput).length === 0) return false;
+  return Object.values(propOutput).every((keywordResults) =>
+    Object.values(keywordResults).every((v) => v === true)
+  );
+};
+
```
Comment (jdesrosiers) [inline src/error-handlers/anyOf.js]: I prefer to put helper functions at the bottom.

File: src/error-handlers/anyOf.js [hunk 2]
```diff
@@ -2,45 +2,88 @@ import * as Instance from "@hyperjump/json-schema/instance/experimental";
...
   /** @type ErrorObject[] */
   const errors = [];
 
-  for (const schemaLocation in normalizedErrors["https://json-schema.org/keyword/anyOf"]) {
-    const anyOf = normalizedErrors["https://json-schema.org/keyword/anyOf"][schemaLocation];
+  for (const schemaLocation in normalizedErrors[
+    "https://json-schema.org/keyword/anyOf"
+  ]) {
+    const anyOf
+      = normalizedErrors["https://json-schema.org/keyword/anyOf"][schemaLocation];
```
Comment (jdesrosiers) [inline src/error-handlers/anyOf.js]: These should not be broken into multiple lines. This does nothing but make the code harder to read. Please try to match the code style used in the project.

File: src/error-handlers/anyOf.js [hunk 3]
```diff
@@ -2,45 +2,88 @@ import * as Instance from "@hyperjump/json-schema/instance/experimental";
...
+        [...Instance.keys(instance)].map(
+          (keyNode) => /** @type {string} */ (Instance.value(keyNode))
+        )
+      );
+      const prefix = `${instanceLocation}/`;
+
+      filtered = filtered.filter((alternative) => {
+        const declaredProps = Object.keys(alternative)
+          .filter((loc) => loc.startsWith(prefix))
+          .map((loc) => loc.slice(prefix.length));
```
Comment (jdesrosiers) [inline src/error-handlers/anyOf.js]: This isn't going to to always work. Consider the case where the pointer is escaped like `~0a`. You'll need to unescape it to get the actual property name.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
Please try to match the code style used in the rest of the project. Please turn off Prettier or at least give it a much more reasonable line length limit.

I think this works! But, getting it to work is just the first step. This needs some refactoring.

Every time you do a filter operation, you're looping over everything again and creating a new array. Creating a bunch of arrays and objects gets expensive at scale. See if you can refactor so that you can loop over the alternatives only one time.

Also try to avoid things like `[...instanceProps]`. Again, we're unnecessarily creating an array. Have a look at `@hyperjump/pact`. It allows you to use `filter`, `map` and bunch of other things that work on any iterator (not just arrays) and don't create unnecessary intermediate representations. It looks like it could come in handy in a few places.

File: src/error-handlers/anyOf.js [hunk 4]
```diff
@@ -15,32 +17,65 @@ const anyOfErrorHandler = async (normalizedErrors, instance, localization) => {
...
+      }
+    } else {
+      filtered = [];
       for (const alternative of anyOf) {
-        alternatives.push(await getErrors(alternative, instance, localization));
+        const typeResults = alternative[instanceLocation]["https://json-schema.org/keyword/type"];
+        if (!typeResults || Object.values(typeResults).every((isValid) => isValid)) {
+          filtered.push(alternative);
+        }
       }
```
Comment (jdesrosiers) [inline src/error-handlers/anyOf.js]: This is now duplicated code. You don't need the `if`/`else`. Remove the `if` wrapper and the `else` block entirely.

**Comment by animeshsahoo1** [comment]
Hi I have fixed the failing npm run lint command with new oxlint linting, please check if its ok.

File: src/error-handlers/anyOf.js [hunk 5]
```diff
@@ -49,4 +84,12 @@ const anyOfErrorHandler = async (normalizedErrors, instance, localization) => {
   return errors;
 };
 
+/** @type (propOutput: NormalizedOutput[string] | undefined) => boolean */
+const propertyPasses = (propOutput) => {
+  if (!propOutput || Object.keys(propOutput).length === 0) {
```
Comment (jdesrosiers) [inline src/error-handlers/anyOf.js]: Checking if the object is empty is unnecessary.

Comment (jdesrosiers) [inline src/error-handlers/anyOf.js]: I don't understand the purpose of the check for an empty object and the tests don't cover it. Either remove it or add a test showing it's needed.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]

File: src/error-handlers/anyOf.js (hunk 4 above)
Comment (animeshsahoo1) [reply src/error-handlers/anyOf.js]: So basically remove if else wrapper put the type check at top for each alternative then if its an object we check rule1 and rule2 right

**Review by jdesrosiers — COMMENTED** [review]
I found a bug. Here's a test that exposes it.

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

  "anyOf": [
    {
      "type": "object",
      "properties": {
        "a": { "type": "string" }
      },
      "required": ["a"]
    },
    {
      "type": "object",
      "properties": {
        "b": { "type": "string" }
      },
      "required": ["b"]
    }
  ]
}
```
```json
{ "a": 42 }
```

The first alternative is getting filtered by the discriminator check (rule 2), but "a" is not a discriminator because it doesn't have a passing value in any other alternatives.

**Comment by animeshsahoo1** [comment]
yeah this is definitely an edge case I was thinking we should add to rule 2:  only apply this filter if the failing properties all have passing constraints in at least one other alternative
this works on the previous example as well as the current example:
```
{
  "anyOf": [
    {
      "type": "object",
      "properties": {
        "type": { "const": "a" },
        "x": { "type": "string" }
      },
      "required": ["type", "x"]
    },
    {
      "type": "object",
      "properties": {
        "type": {
          "allOf": [
            { "type": "string" },
            { "minLength": 2 }
          ]
        },
        "x": { "type": "number" }
      },
      "required": ["type", "x"]
    }
  ]
}
```
`Instance: { "type": "a", "x": 42 }`
here both don't get filtered as suggested earlier and
```
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

  "anyOf": [
    {
      "type": "object",
      "properties": {
        "a": { "type": "string" }
      },
      "required": ["a"]
    },
    {
      "type": "object",
      "properties": {
        "b": { "type": "string" }
      },
      "required": ["b"]
    }
  ]
}
```
`instance: { "a": 42 }
`this also works

so the new rule2 is:
Filter out an alternative if no instance property passes in it, AND all of its failing properties have a passing constraint in at least one other alternative. I wanna look at its edge case a bit more would love your views too.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
This has gotten a bit messy again. I see the same problem as before where you're looping over things multiple times. I don't think we can get away with looping over the alternatives less than three times, but we certainly shouldn't be doing it more than that. I expect one loop to determine discriminators, another to filter alternatives, and a third if all the alternatives got filtered out. Right now you have two loops for filtering.

Also, I see that you removed the use of `Pact.pipe`. I found that considerably harder to read.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
I pushed two commits showing what I was looking for. It might be instructive to study each commit individually. (I only made the changes for `anyOf`, you'll need to do something similar for `oneOf`.

**Review by jdesrosiers — APPROVED** [review]
I was able to simplify the solution even more! There only ended up being one place where Pact was needed. I'm pretty happy with how the code ended up considering how complicated it was to get there. I also did a little cleanup and added a few more tests.

Thanks for your effort on this. I know it was a hard one.

**Comment by animeshsahoo1** [comment]
Sorry it took so much effort the logic itself was known but matching the code style was a bit hard, but I learnt a lot of how you prefer the code to be it will definitely be helpful for future PRs.
